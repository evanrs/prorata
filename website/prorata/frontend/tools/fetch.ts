import { fetch, TypedResponse } from '@evanrs/fetch'

export interface ExtendedRequestInit<T> extends RequestInit {
  data?: T
}

export interface ExtendedResponse<T> extends TypedResponse<T> {
  data: T
}

async function __fetch<D, T>(
  request: RequestInfo,
  { data, headers, body, ...options }: ExtendedRequestInit<D> = {},
): Promise<ExtendedResponse<T>> {
  if (data != null) {
    headers = { ...headers, 'content-type': 'application/json' }
    body = JSON.stringify(data)
  }

  const response = await fetch<T>(
    `${process.env.API_BASE_URL ?? ''}${(request as Request)?.url ?? request}`,
    {
      ...options,
      headers,
      body,
    },
  )

  return Object.assign(response, {
    data: await response.json(),
  })
}

function isRequest(request: RequestInfo): request is Request {
  return String(request) === '[object Request]'
}

export { __fetch as fetch }
