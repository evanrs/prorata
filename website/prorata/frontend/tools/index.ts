export * from './fetch'
export * from './storage'
export * from './is-promise'
export * from './is-equal'
export * from './opacity-transition-for'
