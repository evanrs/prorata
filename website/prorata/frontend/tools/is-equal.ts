export function isEqual<T extends Record<string, unknown>>(
  reference?: T,
  partial?: Partial<T>,
): partial is T {
  if (!reference || !partial) {
    return false
  }

  return Object.keys(reference).every((v, k) => v === partial[k])
}

export function isNotEqual<T extends Record<string, unknown>>(reference?: T, partial?: Partial<T>) {
  return !isEqual(reference, partial)
}
