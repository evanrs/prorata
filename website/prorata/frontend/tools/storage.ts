export function Storage<T>(prefix?: string) {
  const keyFor = (key?: string) => [prefix, key].filter(Boolean).join(':')
  return {
    get(key?: string): T | undefined {
      try {
        return JSON.parse(localStorage.getItem(keyFor(key)) ?? '') ?? undefined
      } catch (error) {
        // fail silently, because that happens sometimes
      }
    },

    set<V extends Partial<T>>(key: string, value: V): V {
      localStorage.setItem(keyFor(key), JSON.stringify(value))
      return value
    },

    clear(key: string) {
      localStorage.removeItem(keyFor(key))
    },
  }
}
