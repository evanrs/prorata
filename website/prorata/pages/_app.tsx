import 'ress/ress.css'
import '@fontsource/inter'
import '@fontsource/nunito-sans'

import type { AppProps } from 'next/app'
import { ChakraProvider, extendTheme, theme } from '@chakra-ui/react'

const extendedTheme = extendTheme({
  fonts: {
    // heading: `inter, ${theme.fonts.heading}`,
    body: `Nunito Sans, ${theme.fonts.body}`,
  },
  config: {
    useSystemColorMode: true,
  },
})

function App({ Component, pageProps }: AppProps) {
  return (
    <ChakraProvider theme={extendedTheme}>
      <Component {...pageProps} />{' '}
    </ChakraProvider>
  )
}

export default App
