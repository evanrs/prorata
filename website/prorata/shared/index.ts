export * from './schema'
export * from './ajv'
export * from './guard-for'
