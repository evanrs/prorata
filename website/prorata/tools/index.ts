export * from './fetch'
export * from './use-call-effect'
