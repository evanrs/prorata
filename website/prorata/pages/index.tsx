import { useState } from 'react'
import { NextPage } from 'next'
import Head from 'next/head'
import { useRouter } from 'next/router'
import absoluteUrl from 'next-absolute-url'

import { AllocationRequest, AllocationResponse } from '../shared'
import { fetch, useAsyncState } from '../client'
import { Prorata } from '../components'

export type HomeProps = Record<string, unknown>

export const Home: NextPage<HomeProps> = (_props) => {
  const [input, setInput] = useState<AllocationRequest>()
  const [output] = useAsyncState<AllocationResponse>(() => {
    if (input == null) {
      return { allocations: [] }
    }

    return fetch<AllocationRequest, AllocationResponse>('/api/prorate', {
      method: 'POST',
      data: input,
    })
      .then((response) => response.data)
      .catch((e) => console.error(e))
  }, [input])

  return <Prorata allocation={output} allocationFor={setInput}></Prorata>
}

Home.getInitialProps = async (context) => {
  console.log('getInitialProps', context)
  return { hello: true }
}

export default Home
