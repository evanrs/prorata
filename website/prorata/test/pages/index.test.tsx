import React from 'react'
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import '@testing-library/jest-dom/extend-expect'

import { AllocationRequest, AllocationResponse } from '../../shared'
import { Home } from '../../pages'
import { render, fetch, fireEvent, act, waitFor, screen } from '../tools'
import { data } from '../data'

describe('Home page', () => {
  it('matches snapshot', async () => {
    const result = render(<Home />, {})
    await waitFor(() => screen.getByText('Investor A'))

    expect(result.asFragment()).toMatchSnapshot()
  })

  xit('clicking button triggers alert', () => {
    // const { getByText } = render(<Home />, {})
    // window.alert = jest.fn()
    // fireEvent.click(getByText('Test Button'))
    // expect(window.alert).toHaveBeenCalledWith('With typescript and Jest')
  })
})
