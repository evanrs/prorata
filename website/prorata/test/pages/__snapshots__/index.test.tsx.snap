// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-z8qy36"
  >
    <h2
      class="chakra-heading css-0"
    >
      prorata
    </h2>
    <div
      class="css-3gx020"
    >
      <a
        class="chakra-button css-5rlsx0"
        href="mailto:me@evanrs.com?subject=Evan%2C%20nice%20work%20on%20Prorata%20%F0%9F%91%8F&body=Hey%20Evan%2C%0A%0AI%20just%20wanted%20say%20%E2%80%A6&cc=evanrs@gmail.com"
      >
        <svg
          class="chakra-icon css-onkibi"
          focusable="false"
          viewBox="0 0 14 14"
        >
          <path
            d="M0.913134,0.920639 C1.49851,0.331726 2.29348,0 3.12342,0 L10.8766,0 C11.7065,0 12.5015,0.331725 13.0869,0.920639 C13.6721,1.50939 14,2.30689 14,3.13746 L14,8.12943 C13.9962,8.51443 13.9059,8.97125 13.7629,9.32852 C13.6128,9.683 13.3552,10.0709 13.0869,10.3462 C12.813,10.6163 12.4265,10.8761 12.0734,11.0274 C11.7172,11.1716 11.2607,11.263 10.8766,11.2669 L10.1234,11.2669 L10.1234,12.5676 L10.1209,12.5676 C10.1204,12.793 10.0633,13.0791 9.97807,13.262 C9.8627,13.466 9.61158,13.7198 9.40818,13.8382 L9.40824,13.8383 C9.4077,13.8386 9.40716,13.8388 9.40661,13.8391 C9.40621,13.8393 9.4058,13.8396 9.40539,13.8398 L9.40535,13.8397 C9.22958,13.9254 8.94505,13.9951 8.75059,14 L8.74789,14 C8.35724,13.9963 7.98473,13.8383 7.71035,13.5617 L5.39553,11.2669 L3.12342,11.2669 C2.29348,11.2669 1.49851,10.9352 0.913134,10.3462 C0.644826,10.0709 0.387187,9.683 0.23711,9.32852 C0.0941235,8.97125 0.00379528,8.51443 0,8.12943 L0,3.13746 C0,2.30689 0.327915,1.50939 0.913134,0.920639 Z M3.12342,1.59494 C2.71959,1.59494 2.33133,1.75628 2.04431,2.04503 C1.75713,2.33395 1.59494,2.72681 1.59494,3.13746 L1.59494,8.12943 C1.59114,8.35901 1.62114,8.51076 1.71193,8.72129 C1.79563,8.9346 1.88065,9.06264 2.04431,9.22185 C2.33133,9.5106 2.71959,9.67195 3.12342,9.67195 L5.72383,9.67195 C5.93413,9.67195 6.13592,9.75502 6.28527,9.90308 L8.52848,12.1269 L8.52848,10.4694 C8.52848,10.029 8.88552,9.67195 9.32595,9.67195 L10.8766,9.67195 C11.1034,9.67583 11.2517,9.64614 11.4599,9.55518 C11.6712,9.47132 11.7976,9.38635 11.9557,9.22185 C12.1193,9.06264 12.2044,8.9346 12.2881,8.72129 C12.3789,8.51076 12.4089,8.35901 12.4051,8.12943 L12.4051,3.13746 C12.4051,2.72681 12.2429,2.33394 11.9557,2.04503 C11.6687,1.75628 11.2804,1.59494 10.8766,1.59494 L3.12342,1.59494 Z"
            fill="currentColor"
          />
        </svg>
      </a>
      <button
        class="chakra-button css-5rlsx0"
        type="button"
      >
        <svg
          class="chakra-icon css-1pcfsbz"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M21.4,13.7C20.6,13.9,19.8,14,19,14c-5,0-9-4-9-9c0-0.8,0.1-1.6,0.3-2.4c0.1-0.3,0-0.7-0.3-1 c-0.3-0.3-0.6-0.4-1-0.3C4.3,2.7,1,7.1,1,12c0,6.1,4.9,11,11,11c4.9,0,9.3-3.3,10.6-8.1c0.1-0.3,0-0.7-0.3-1 C22.1,13.7,21.7,13.6,21.4,13.7z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        class="chakra-button css-5rlsx0"
        type="button"
      >
        <svg
          class="chakra-icon css-dvq678"
          focusable="false"
          viewBox="0 0 16 16"
        >
          <path
            d="M9.41 8l2.29-2.29c.19-.18.3-.43.3-.71a1.003 1.003 0 0 0-1.71-.71L8 6.59l-2.29-2.3a1.003 1.003 0 0 0-1.42 1.42L6.59 8 4.3 10.29c-.19.18-.3.43-.3.71a1.003 1.003 0 0 0 1.71.71L8 9.41l2.29 2.29c.18.19.43.3.71.3a1.003 1.003 0 0 0 .71-1.71L9.41 8z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    class="css-apl5pd"
  >
    <a
      class="chakra-button css-9njja2"
      href="mailto:me@evanrs.com?subject=Evan%2C%20nice%20work%20on%20Prorata%20%F0%9F%91%8F&body=Hey%20Evan%2C%0A%0AI%20just%20wanted%20say%20%E2%80%A6&cc=evanrs@gmail.com"
    >
      <svg
        class="chakra-icon surprise css-onkibi"
        focusable="false"
        viewBox="0 0 14 14"
      >
        <path
          d="M0.913134,0.920639 C1.49851,0.331726 2.29348,0 3.12342,0 L10.8766,0 C11.7065,0 12.5015,0.331725 13.0869,0.920639 C13.6721,1.50939 14,2.30689 14,3.13746 L14,8.12943 C13.9962,8.51443 13.9059,8.97125 13.7629,9.32852 C13.6128,9.683 13.3552,10.0709 13.0869,10.3462 C12.813,10.6163 12.4265,10.8761 12.0734,11.0274 C11.7172,11.1716 11.2607,11.263 10.8766,11.2669 L10.1234,11.2669 L10.1234,12.5676 L10.1209,12.5676 C10.1204,12.793 10.0633,13.0791 9.97807,13.262 C9.8627,13.466 9.61158,13.7198 9.40818,13.8382 L9.40824,13.8383 C9.4077,13.8386 9.40716,13.8388 9.40661,13.8391 C9.40621,13.8393 9.4058,13.8396 9.40539,13.8398 L9.40535,13.8397 C9.22958,13.9254 8.94505,13.9951 8.75059,14 L8.74789,14 C8.35724,13.9963 7.98473,13.8383 7.71035,13.5617 L5.39553,11.2669 L3.12342,11.2669 C2.29348,11.2669 1.49851,10.9352 0.913134,10.3462 C0.644826,10.0709 0.387187,9.683 0.23711,9.32852 C0.0941235,8.97125 0.00379528,8.51443 0,8.12943 L0,3.13746 C0,2.30689 0.327915,1.50939 0.913134,0.920639 Z M3.12342,1.59494 C2.71959,1.59494 2.33133,1.75628 2.04431,2.04503 C1.75713,2.33395 1.59494,2.72681 1.59494,3.13746 L1.59494,8.12943 C1.59114,8.35901 1.62114,8.51076 1.71193,8.72129 C1.79563,8.9346 1.88065,9.06264 2.04431,9.22185 C2.33133,9.5106 2.71959,9.67195 3.12342,9.67195 L5.72383,9.67195 C5.93413,9.67195 6.13592,9.75502 6.28527,9.90308 L8.52848,12.1269 L8.52848,10.4694 C8.52848,10.029 8.88552,9.67195 9.32595,9.67195 L10.8766,9.67195 C11.1034,9.67583 11.2517,9.64614 11.4599,9.55518 C11.6712,9.47132 11.7976,9.38635 11.9557,9.22185 C12.1193,9.06264 12.2044,8.9346 12.2881,8.72129 C12.3789,8.51076 12.4089,8.35901 12.4051,8.12943 L12.4051,3.13746 C12.4051,2.72681 12.2429,2.33394 11.9557,2.04503 C11.6687,1.75628 11.2804,1.59494 10.8766,1.59494 L3.12342,1.59494 Z"
          fill="currentColor"
        />
      </svg>
      <div
        class="css-df6gl"
      />
      <h2
        class="chakra-heading css-0"
      >
        Evan Schneider
      </h2>
    </a>
  </div>
  <div
    class="css-h74a88"
  >
    <div
      class="css-1umrmxj"
    >
      <div
        class="css-j7qwjs"
      >
        <div
          class="css-1sfbe0h"
        >
          <h2
            class="chakra-heading css-1jd2nkm"
          >
            Total Available Allocation
          </h2>
          <div />
          <div />
          <h2
            class="chakra-heading css-c1wsrx"
          >
            Total Allocated
          </h2>
        </div>
        <div
          class="css-1qvt70z"
        >
          <div
            class="chakra-numberinput css-e69dqy"
            value=""
          >
            <input
              aria-valuemax="9007199254740991"
              aria-valuemin="1"
              autocomplete="off"
              autocorrect="off"
              autofocus=""
              class="chakra-numberinput__field css-1o8sjxt"
              inputmode="numeric"
              name="allocation_amount"
              pattern="^[+-]?[$]?[0-9]{1,3}(?:,?[0-9]{3})*(?:\\\\.[0-9]{2})?$"
              placeholder="Allocation"
              role="spinbutton"
              type="text"
              value=""
              variant="filled"
            />
          </div>
          <div />
          <div />
          <div
            class="chakra-numberinput css-mpussp"
            value=""
          >
            <input
              aria-readonly="true"
              aria-valuemax="9007199254740991"
              aria-valuemin="-9007199254740991"
              autocomplete="off"
              autocorrect="off"
              class="chakra-numberinput__field css-1o8sjxt"
              inputmode="numeric"
              name="total"
              pattern="^[+-]?[$]?[0-9]{1,3}(?:,?[0-9]{3})*(?:\\\\.[0-9]{2})?$"
              readonly=""
              role="spinbutton"
              type="text"
              value=""
              variant="filled"
            />
          </div>
          <div />
        </div>
        <div
          class="css-11ucw7v"
        >
          <h2
            class="chakra-heading css-msdjfj"
          >
            Investor Breakdown
          </h2>
          <h2
            class="chakra-heading css-msdjfj"
          />
          <h2
            class="chakra-heading css-msdjfj"
          />
          <h2
            class="chakra-heading css-crdmyz"
          >
            Investor Stake
          </h2>
          <div />
        </div>
        <form
          autocomplete="off"
          class="css-1l4fvs0"
        >
          <input
            autocomplete="false"
            hidden=""
            name="hidden"
          />
          <input
            autocomplete="off"
            class="chakra-input css-taf3kg"
            name="name"
            placeholder="Name"
            value=""
          />
          <div
            class="chakra-numberinput css-e69dqy"
            value=""
          >
            <input
              aria-valuemax="9007199254740991"
              aria-valuemin="0"
              autocomplete="off"
              autocorrect="off"
              class="chakra-numberinput__field css-1o8sjxt"
              inputmode="numeric"
              name="requested_amount"
              pattern="^[+-]?[$]?[0-9]{1,3}(?:,?[0-9]{3})*(?:\\\\.[0-9]{2})?$"
              placeholder="Requested Amount"
              role="spinbutton"
              type="text"
              value=""
              variant="filled"
            />
          </div>
          <div
            class="chakra-numberinput css-e69dqy"
            value=""
          >
            <input
              aria-valuemax="9007199254740991"
              aria-valuemin="0"
              autocomplete="off"
              autocorrect="off"
              class="chakra-numberinput__field css-1o8sjxt"
              inputmode="numeric"
              name="average_amount"
              pattern="^[+-]?[$]?[0-9]{1,3}(?:,?[0-9]{3})*(?:\\\\.[0-9]{2})?$"
              placeholder="Average Amount"
              role="spinbutton"
              type="text"
              value=""
              variant="filled"
            />
          </div>
          <div
            class="chakra-numberinput css-3uwy3r"
            value=""
          >
            <input
              aria-readonly="true"
              aria-valuemax="9007199254740991"
              aria-valuemin="-9007199254740991"
              autocomplete="off"
              autocorrect="off"
              class="chakra-numberinput__field css-1o8sjxt"
              inputmode="numeric"
              name="allocation"
              pattern="^[+-]?[$]?[0-9]{1,3}(?:,?[0-9]{3})*(?:\\\\.[0-9]{2})?$"
              placeholder="Stake"
              readonly=""
              role="spinbutton"
              type="text"
              value=""
              variant="filled"
            />
          </div>
          <button
            class="chakra-button css-1xx8uq7"
            disabled=""
            type="submit"
          >
            <svg
              class="chakra-icon css-g7wnnj"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M0,12a1.5,1.5,0,0,0,1.5,1.5h8.75a.25.25,0,0,1,.25.25V22.5a1.5,1.5,0,0,0,3,0V13.75a.25.25,0,0,1,.25-.25H22.5a1.5,1.5,0,0,0,0-3H13.75a.25.25,0,0,1-.25-.25V1.5a1.5,1.5,0,0,0-3,0v8.75a.25.25,0,0,1-.25.25H1.5A1.5,1.5,0,0,0,0,12Z"
                fill="currentColor"
              />
            </svg>
          </button>
          <input
            hidden=""
            type="submit"
            value="Submit"
          />
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
